// @report-designer/core/dist/render-engine/index.js
import { evalExpression } from '../expression-engine';
/** Resolve an expression string or return static value */
export function resolveValue(value, ctx) {
    if (!value)
        return value;
    // If it looks like an expression
    if (value.includes('{') || value.includes('(') || value.includes('=')) {
        try {
            return evalExpression(value, ctx.resolveField.bind(ctx), ctx.rowIndex, ctx.variables);
        }
        catch {
            return value;
        }
    }
    return value;
}
/** Look up a style by ID from the template's style list */
function getStyleById(comp, styles) {
    if (!comp.style)
        return undefined;
    return styles.find(s => s.id === comp.style);
}
/** Convert template component to rendered style */
function componentStyleToRendered(comp, templateStyles) {
    const refStyle = getStyleById(comp, templateStyles);
    const result = {};
    if (refStyle) {
        result.font = refStyle.font;
        result.background = refStyle.backgroundColor;
        result.border = refStyle.border;
    }
    // Component-level overrides
    if (comp.type === 'text') {
        const tc = comp;
        result.textAlign = tc.textAlign;
        result.canGrow = tc.canGrow;
        if (tc.font) {
            result.font = { ...result.font, ...tc.font };
        }
    }
    return result;
}
/** Apply conditional rules to a component's style */
function applyConditionalRules(baseStyle, formats, comp, ctx) {
    let style = { ...baseStyle };
    for (const fmt of formats) {
        // Check if this format applies to this component
        if (fmt.applyTo.length > 0 && !fmt.applyTo.includes(comp.id)) {
            continue;
        }
        for (const rule of fmt.rules) {
            if (rule.expression) {
                try {
                    const result = evalExpression(rule.expression, ctx.resolveField.bind(ctx), ctx.rowIndex, ctx.variables);
                    if (result) {
                        // Apply overrides to style
                        if (rule.overrides.font) {
                            style.font = { ...style.font, ...rule.overrides.font };
                        }
                        if (rule.overrides.backgroundColor !== undefined) {
                            style.background = rule.overrides.backgroundColor;
                        }
                        if (rule.overrides.border) {
                            style.border = { ...style.border, ...rule.overrides.border };
                        }
                        if (rule.overrides.textAlign) {
                            style.textAlign = rule.overrides.textAlign;
                        }
                    }
                }
                catch {
                    // Skip failed condition evaluations
                }
            }
        }
    }
    return style;
}
/** Render a single component with expression resolution */
function renderComponent(comp, ctx, formats, templateStyles) {
    const baseStyle = componentStyleToRendered(comp, templateStyles);
    const style = applyConditionalRules(baseStyle, formats, comp, ctx);
    const rendered = {
        id: comp.id,
        type: comp.type,
        x: comp.x,
        y: comp.y,
        width: comp.width,
        height: comp.height,
        style,
    };
    if (comp.type === 'text') {
        const textComp = comp;
        const resolved = resolveValue(textComp.text, ctx);
        rendered.content = resolved != null ? String(resolved) : '';
    }
    if (comp.type === 'image') {
        const imgComp = comp;
        const resolved = resolveValue(imgComp.src, ctx);
        rendered.imageSrc = resolved ? String(resolved) : '';
    }
    if (comp.type === 'barcode') {
        const bcComp = comp;
        const resolved = resolveValue(bcComp.value, ctx);
        rendered.barcodeValue = resolved ? String(resolved) : '';
    }
    return rendered;
}
/** Render a single band */
function renderBand(band, ctx, formats, templateStyles) {
    const components = band.components.map(c => renderComponent(c, ctx, formats, templateStyles));
    return {
        id: band.id,
        type: band.type,
        height: band.height,
        components,
        isDataBand: band.type === 'data',
        isGroupHeader: band.type === 'groupHeader',
        isGroupFooter: band.type === 'groupFooter',
    };
}
/** Main renderer: produces a RenderTree from template + data */
export function renderTemplate(template, data) {
    const formats = template.conditionalFormats || [];
    const templateStyles = template.styles || [];
    const pages = [];
    const rowCount = {};
    for (const page of template.pages) {
        const renderedBands = [];
        for (const band of page.bands) {
            // Static bands (title, header, footer) - render once with no data context
            if (band.type === 'reportTitle' || band.type === 'pageHeader' || band.type === 'pageFooter') {
                const ctx = {
                    resolveField: () => null,
                    rowIndex: 0,
                    variables: {},
                };
                renderedBands.push(renderBand(band, ctx, formats, templateStyles));
                continue;
            }
            // Data band - render per row
            if (band.type === 'data' && band.dataSource) {
                const rows = data[band.dataSource] || [];
                rowCount[band.dataSource] = (rowCount[band.dataSource] || 0) + rows.length;
                for (let i = 0; i < rows.length; i++) {
                    const row = rows[i];
                    const ctx = {
                        resolveField: (source, field) => {
                            if (source && row[source] && typeof row[source] === 'object') {
                                return row[source][field];
                            }
                            return row[field] ?? row[source] ?? null;
                        },
                        rowIndex: i,
                        variables: { rowIndex: i, row: row },
                    };
                    renderedBands.push(renderBand(band, ctx, formats, templateStyles));
                }
                continue;
            }
            // Default: render once
            const ctx = {
                resolveField: () => null,
                rowIndex: 0,
                variables: {},
            };
            renderedBands.push(renderBand(band, ctx, formats, templateStyles));
        }
        pages.push({
            id: page.id,
            bands: renderedBands,
        });
    }
    return {
        pages,
        templateId: template.id,
        totalPages: pages.length,
        rowCount,
    };
}
/** Get a summary of the render tree for debugging */
export function renderTreeSummary(tree) {
    const lines = [];
    for (const page of tree.pages) {
        lines.push(`Page ${page.id}:`);
        for (const band of page.bands) {
            lines.push(`  Band ${band.id} (${band.type}): ${band.components.length} components`);
        }
    }
    return lines.join('\n');
}
