// @report-designer/core/dist/expression-engine/parser.js
import { TokenType } from './lexer';
import { ASTNodeType, } from './ast';
export class ParseError extends Error {
    token;
    constructor(message, token) {
        super(message);
        this.token = token;
    }
}
export function parse(tokens) {
    const parser = new Parser(tokens);
    const node = parser.parseExpression();
    if (parser.isEOF()) {
        const t = parser.lastToken();
        throw new ParseError(`Unexpected token: ${t.value}`, t);
    }
    return node;
}
class Parser {
    pos = 0;
    tokens;
    constructor(tokens) {
        this.tokens = tokens;
    }
    current() {
        return this.tokens[this.pos];
    }
    peek(offset = 0) {
        return this.tokens[this.pos + offset];
    }
    isEOF() {
        return this.tokens[this.pos].type !== TokenType.EOF;
    }
    lastToken() {
        return this.tokens[this.pos];
    }
    consume(expected) {
        const t = this.current();
        if (t.type !== expected) {
            throw new ParseError(`Expected token type ${TokenType[expected]} but got ${TokenType[t.type]} (${t.value})`, t);
        }
        this.pos++;
        return t;
    }
    /**
     * Expression = OrExpr
     * Lowest precedence: OR
     */
    parseExpression() {
        return this.parseOr();
    }
    /** OR: additive ( OR additive )* */
    parseOr() {
        let node = this.parseAnd();
        while (this.current().type === TokenType.OR) {
            const op = this.current().value;
            this.pos++;
            const right = this.parseAnd();
            node = this.makeBinary(op, node, right);
        }
        return node;
    }
    /** AND: comparison ( AND comparison )* */
    parseAnd() {
        let node = this.parseComparison();
        while (this.current().type === TokenType.AND) {
            const op = this.current().value;
            this.pos++;
            const right = this.parseComparison();
            node = this.makeBinary(op, node, right);
        }
        return node;
    }
    /** Comparison: additive ((= | != | > | < | >= | <=) additive)* */
    parseComparison() {
        let node = this.parseAdditive();
        const compTypes = [
            TokenType.EQUALS, TokenType.NOT_EQUALS,
            TokenType.GT, TokenType.LT,
            TokenType.GTE, TokenType.LTE,
        ];
        while (compTypes.includes(this.current().type)) {
            const op = this.current().value;
            this.pos++;
            const right = this.parseAdditive();
            node = this.makeBinary(op, node, right);
        }
        return node;
    }
    /** Additive: multiplicative ((+ | -) multiplicative)* */
    parseAdditive() {
        let node = this.parseMultiplicative();
        while (this.current().type === TokenType.PLUS || this.current().type === TokenType.MINUS) {
            const op = this.current().value;
            this.pos++;
            const right = this.parseMultiplicative();
            node = this.makeBinary(op, node, right);
        }
        return node;
    }
    /** Multiplicative: unary ((* | / | %) unary)* */
    parseMultiplicative() {
        let node = this.parseUnary();
        while ([TokenType.MULTIPLY, TokenType.DIVIDE, TokenType.MODULO].includes(this.current().type)) {
            const op = this.current().value;
            this.pos++;
            const right = this.parseUnary();
            node = this.makeBinary(op, node, right);
        }
        return node;
    }
    /** Unary: NOT unary | MINUS unary | primary */
    parseUnary() {
        if (this.current().type === TokenType.NOT) {
            const op = this.current().value;
            this.pos++;
            const operand = this.parseUnary();
            return this.makeUnary(op, operand);
        }
        if (this.current().type === TokenType.MINUS) {
            const op = this.current().value;
            this.pos++;
            const operand = this.parseUnary();
            return this.makeUnary(op, operand);
        }
        return this.parsePrimary();
    }
    /** Primary: LPAREN expression RPAREN | IDENTIFIER LPAREN args RPAREN | FIELD_REF | STRING | NUMBER | BOOLEAN */
    parsePrimary() {
        const t = this.current();
        // Parenthesized expression
        if (t.type === TokenType.LPAREN) {
            this.pos++;
            const node = this.parseExpression();
            this.consume(TokenType.RPAREN);
            return node;
        }
        // Function call: IDENTIFIER LPAREN
        if (t.type === TokenType.IDENTIFIER && this.peek(1).type === TokenType.LPAREN) {
            this.pos++; // consume identifier
            this.pos++; // consume LPAREN
            const args = this.parseArgs();
            this.consume(TokenType.RPAREN);
            return this.makeFunctionCall(t.value, args);
        }
        // Field reference
        if (t.type === TokenType.FIELD_REF) {
            this.pos++;
            return this.makeFieldRef(t.value);
        }
        // String literal
        if (t.type === TokenType.STRING) {
            this.pos++;
            return this.makeStringLiteral(t.value);
        }
        // Number literal
        if (t.type === TokenType.NUMBER) {
            this.pos++;
            return this.makeNumberLiteral(t.value);
        }
        // Boolean literal
        if (t.type === TokenType.BOOLEAN) {
            this.pos++;
            return this.makeBooleanLiteral(t.value);
        }
        // Null literal
        if (t.type === TokenType.NULL) {
            this.pos++;
            return this.makeNullLiteral();
        }
        // Bare identifier — treat as a function with no args (e.g. a variable reference)
        if (t.type === TokenType.IDENTIFIER) {
            this.pos++;
            // For now, treat bare identifiers as function calls with no args.
            // This matches Stimulsoft's behavior where field refs in expressions
            // are typically wrapped in {}, but bare names can resolve to variables.
            return this.makeFunctionCall(t.value, []);
        }
        throw new ParseError(`Unexpected token: ${TokenType[t.type]} (${t.value})`, t);
    }
    /** Parse comma-separated argument list inside parentheses */
    parseArgs() {
        if (this.current().type === TokenType.RPAREN) {
            return [];
        }
        const args = [];
        args.push(this.parseExpression());
        while (this.current().type === TokenType.COMMA) {
            this.pos++;
            args.push(this.parseExpression());
        }
        return args;
    }
    // ---- Node factories ----
    makeBinary(op, left, right) {
        return { type: ASTNodeType.BinaryOp, operator: op, left, right };
    }
    makeUnary(op, operand) {
        return { type: ASTNodeType.UnaryOp, operator: op, operand };
    }
    makeFieldRef(value) {
        const parts = value.split('.');
        const source = parts.length > 1 ? parts.slice(0, -1).join('.') : '';
        const field = parts.length > 0 ? parts[parts.length - 1] : value;
        return { type: ASTNodeType.FieldRef, source, field };
    }
    makeStringLiteral(value) {
        return { type: ASTNodeType.Literal, value, dataType: 'string' };
    }
    makeNumberLiteral(value) {
        return { type: ASTNodeType.Literal, value: parseFloat(value), dataType: 'number' };
    }
    makeBooleanLiteral(value) {
        return { type: ASTNodeType.Literal, value: value === 'true', dataType: 'boolean' };
    }
    makeNullLiteral() {
        return { type: ASTNodeType.Literal, value: null, dataType: 'null' };
    }
    makeFunctionCall(name, args) {
        return { type: ASTNodeType.FunctionCall, name, args };
    }
}
