// @report-designer/core/dist/command-engine/index.js
/** Whitelisted commands that are allowed to execute */
const allowedCommands = new Set();
/** Register a command type with its execute and undo handlers */
export function registerCommand(type, execute, undo) {
    allowedCommands.add(type);
    commandHandlers[type] = execute;
    undoHandlers[type] = undo;
}
const commandHandlers = {};
const undoHandlers = {};
/** Command execution error */
export class CommandError extends Error {
    commandType;
    constructor(message, commandType) {
        super(message);
        this.commandType = commandType;
    }
}
/**
 * CommandDispatcher with whitelist security.
 * Only registered commands can be executed.
 */
export class CommandDispatcher {
    history = [];
    future = [];
    maxHistory = 100;
    /** Check if a command type is allowed */
    static isAllowed(type) {
        return allowedCommands.has(type);
    }
    /** List all registered command types */
    static listCommands() {
        return Array.from(allowedCommands);
    }
    /** Execute a command */
    execute(state, command) {
        if (!CommandDispatcher.isAllowed(command.type)) {
            throw new CommandError(`Command '${command.type}' is not whitelisted`, command.type);
        }
        const handler = commandHandlers[command.type];
        if (!handler) {
            throw new CommandError(`No handler registered for '${command.type}'`, command.type);
        }
        // Execute the command
        const newState = handler(state, command.payload);
        // Record in history
        this.history.push(command);
        this.future = []; // Clear redo stack
        // Limit history size
        if (this.history.length > this.maxHistory) {
            this.history = this.history.slice(-this.maxHistory);
        }
        return newState;
    }
    /** Undo the last command */
    undo(state) {
        if (this.history.length === 0)
            return null;
        const command = this.history.pop();
        const handler = undoHandlers[command.type];
        if (!handler) {
            throw new CommandError(`No undo handler for '${command.type}'`, command.type);
        }
        const newState = handler(state, command.payload);
        this.future.push(command);
        return newState;
    }
    /** Redo the last undone command */
    redo(state) {
        if (this.future.length === 0)
            return null;
        const command = this.future.pop();
        const handler = commandHandlers[command.type];
        if (!handler) {
            throw new CommandError(`No handler registered for '${command.type}'`, command.type);
        }
        const newState = handler(state, command.payload);
        this.history.push(command);
        return newState;
    }
    /** Check if undo is available */
    canUndo() {
        return this.history.length > 0;
    }
    /** Check if redo is available */
    canRedo() {
        return this.future.length > 0;
    }
    /** Get the current history size */
    get historySize() {
        return this.history.length;
    }
    /** Clear history */
    clear() {
        this.history = [];
        this.future = [];
    }
}
// ---- Built-in Command Handlers ----
/** Add a component to a band */
registerCommand('add-component', (state, payload) => {
    const { pageId, bandId, component } = payload;
    return {
        ...state,
        pages: state.pages.map(p => {
            if (p.id !== pageId)
                return p;
            return {
                ...p,
                bands: p.bands.map(b => {
                    if (b.id !== bandId)
                        return b;
                    return { ...b, components: [...b.components, component] };
                }),
            };
        }),
    };
}, (state, payload) => {
    const { pageId, bandId, component } = payload;
    return {
        ...state,
        pages: state.pages.map(p => {
            if (p.id !== pageId)
                return p;
            return {
                ...p,
                bands: p.bands.map(b => {
                    if (b.id !== bandId)
                        return b;
                    return { ...b, components: b.components.filter(c => c.id !== component.id) };
                }),
            };
        }),
    };
});
/** Remove a component from a band */
registerCommand('remove-component', (state, payload) => {
    const { pageId, bandId, componentId } = payload;
    return {
        ...state,
        pages: state.pages.map(p => {
            if (p.id !== pageId)
                return p;
            return {
                ...p,
                bands: p.bands.map(b => {
                    if (b.id !== bandId)
                        return b;
                    return { ...b, components: b.components.filter(c => c.id !== componentId) };
                }),
            };
        }),
    };
}, (state, payload) => {
    // Undo of remove = re-add (but we need the component data)
    // The payload should include the removed component
    const { pageId, bandId, component } = payload;
    return {
        ...state,
        pages: state.pages.map(p => {
            if (p.id !== pageId)
                return p;
            return {
                ...p,
                bands: p.bands.map(b => {
                    if (b.id !== bandId)
                        return b;
                    return { ...b, components: [...b.components, component] };
                }),
            };
        }),
    };
});
/** Update a component's properties */
registerCommand('update-component', (state, payload) => {
    const { pageId, bandId, componentId, updates } = payload;
    return {
        ...state,
        pages: state.pages.map(p => {
            if (p.id !== pageId)
                return p;
            return {
                ...p,
                bands: p.bands.map(b => {
                    if (b.id !== bandId)
                        return b;
                    return {
                        ...b,
                        components: b.components.map(c => {
                            if (c.id !== componentId)
                                return c;
                            return { ...c, ...updates };
                        }),
                    };
                }),
            };
        }),
    };
}, (state, payload) => {
    const { pageId, bandId, componentId, previous } = payload;
    if (!previous)
        return state;
    return {
        ...state,
        pages: state.pages.map(p => {
            if (p.id !== pageId)
                return p;
            return {
                ...p,
                bands: p.bands.map(b => {
                    if (b.id !== bandId)
                        return b;
                    return {
                        ...b,
                        components: b.components.map(c => {
                            if (c.id !== componentId)
                                return c;
                            return { ...c, ...previous };
                        }),
                    };
                }),
            };
        }),
    };
});
/** Update a band's height */
registerCommand('resize-band', (state, payload) => {
    const { pageId, bandId, newHeight } = payload;
    return {
        ...state,
        pages: state.pages.map(p => {
            if (p.id !== pageId)
                return p;
            return {
                ...p,
                bands: p.bands.map(b => {
                    if (b.id !== bandId)
                        return b;
                    return { ...b, height: newHeight };
                }),
            };
        }),
    };
}, (state, payload) => {
    const { pageId, bandId, oldHeight } = payload;
    if (oldHeight === undefined)
        return state;
    return {
        ...state,
        pages: state.pages.map(p => {
            if (p.id !== pageId)
                return p;
            return {
                ...p,
                bands: p.bands.map(b => {
                    if (b.id !== bandId)
                        return b;
                    return { ...b, height: oldHeight };
                }),
            };
        }),
    };
});
/** Move a component (update x, y) */
registerCommand('move-component', (state, payload) => {
    const { pageId, bandId, componentId, x, y } = payload;
    return {
        ...state,
        pages: state.pages.map(p => {
            if (p.id !== pageId)
                return p;
            return {
                ...p,
                bands: p.bands.map(b => {
                    if (b.id !== bandId)
                        return b;
                    return {
                        ...b,
                        components: b.components.map(c => {
                            if (c.id !== componentId)
                                return c;
                            return { ...c, x: x ?? c.x, y: y ?? c.y };
                        }),
                    };
                }),
            };
        }),
    };
}, (state, payload) => {
    const { pageId, bandId, componentId, prevX, prevY } = payload;
    if (prevX === undefined && prevY === undefined)
        return state;
    return {
        ...state,
        pages: state.pages.map(p => {
            if (p.id !== pageId)
                return p;
            return {
                ...p,
                bands: p.bands.map(b => {
                    if (b.id !== bandId)
                        return b;
                    return {
                        ...b,
                        components: b.components.map(c => {
                            if (c.id !== componentId)
                                return c;
                            return { ...c, x: prevX ?? c.x, y: prevY ?? c.y };
                        }),
                    };
                }),
            };
        }),
    };
});
/** Set template property */
registerCommand('set-template', (state, payload) => {
    return { ...state, ...payload };
}, (state, payload) => {
    const { previous } = payload;
    if (!previous)
        return state;
    return { ...state, ...previous };
});
